packages/compass-data-modeling/src/components/export-diagram-modal.tsx

```diff
@@ -59,20 +63,71 @@ const ExportDiagramModal = ({
   model,
   onCloseClick,
 }: ExportDiagramModalProps) => {
-  const [exportFormat, setExportFormat] = useState<'json' | null>(null);
-
-  const onExport = useCallback(() => {
-    if (!exportFormat || !model) {
-      return;
+  const [exportFormat, setExportFormat] = useState<'png' | 'json' | null>(null);
+  const diagram = useDiagram();
+  const [isExporting, setIsExporting] = useState(false);
+  const abortControllerRef = useRef<AbortController | null>(null);
+  const toast = useToast();
+  useEffect(() => {
+    const cleanup = () => {
+      if (abortControllerRef.current) {
+        abortControllerRef.current.abort();
+        abortControllerRef.current = null;
+      }
+    };
+    const abortController = new AbortController();
+    if (isModalOpen) {
+      abortControllerRef.current = abortController;
+    } else {
+      cleanup();
     }
-    exportToJson(diagramLabel, model);
+    return cleanup;
+  }, [isModalOpen]);
+
+  const onClose = useCallback(() => {
+    setExportFormat(null);
+    setIsExporting(false);
+    abortControllerRef.current?.abort();
+    abortControllerRef.current = null;
     onCloseClick();
-  }, [exportFormat, onCloseClick, model, diagramLabel]);
+  }, [onCloseClick]);
+
+  const onExport = useCallback(async () => {
```

Comment:
There is way too much logic in the UI here (like if you see that you're keeping abort controllers in render, you probably should start thinking about that), this should really be an action in the store instead

Reply:
Yeah, i'll add another slice for export. Or, I can do it as a follow up if that sounds okay to you.

Reply:
Follow-up sounds good, let's not hold this code in the branch for too long 👍